File: Day7Part1.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.stream.Collectors;
import java.util.Collections;
import java.util.Scanner;
import java.util.regex.MatchResult;
import java.util.List;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Day7Part1 {
    public static void main(String[] args) throws Exception {

        Map<String,List<String>> rules = Files.lines(Paths.get("input-day7.txt"))
            .map(Day7Part1::parseRule)
            .collect(Collectors.toMap(Bag::getColor, Bag::getContainedColors));

        long count = rules.entrySet().stream().map(entry -> containsShinyGold(entry.getKey(), rules))
            .filter(Optional::isPresent)
            .count();

        System.out.println(count);
    }

    static Optional<String> containsShinyGold(String color, Map<String,List<String>> rules) {
        List<String> containedColors = rules.get(color);
        return containedColors.contains("shiny gold") 
            ? Optional.of(color) 
            : containedColors.stream()
                .map(rule -> containsShinyGold(rule, rules))
                .filter(Optional::isPresent)
                .findFirst()
                .orElse(Optional.empty());
    }


    static class Bag {
        String color;
        List<String> containedColors;

        Bag(String color) {
            this(color, Collections.emptyList());
        }

        Bag(String color, List<String> containedColors) {
            this.color = color;
            this.containedColors = containedColors;
        }

        String getColor() { return color; }
        List<String> getContainedColors() { return containedColors; }

        public String toString() {
            return String.format("Color: %s, containedColors: %s", color, containedColors);
        }
    }

    static Bag parseRule(String s) { 
        Scanner scanner = new Scanner(s);
        scanner.findInLine("(.+) bags contain ((\\d+ .+ bags?(,|\\.))|no other bags)");
        MatchResult mr = scanner.match();
        String color = mr.group(1);
        String rulesStr = mr.group(2);
        if (rulesStr.equals("no other bags")) return new Bag(color);

        List<String> containedColors = Arrays.stream(rulesStr.split(", ")).map(r -> {
            String[] tokens = r.split(" ");
            int count = Integer.parseInt(tokens[0]);
            String containedColor = tokens[1] + " " + tokens[2];
            return containedColor;
        })
        .collect(Collectors.toList());

        return new Bag(color, containedColors);
    }
}

File: day7part1.clj
(defn tokenize [line] (.split line " "))

(defn lines [filename] (.split (slurp filename) "\n"))

(defn bag
  [color cnt contained]
  {:color color :count cnt :contained contained})

(defn make-color [tokens index]
   (str (nth tokens index) " " (nth tokens (inc index))))

(defn make-contained-bags
  [tokens]
  (loop [tokens tokens bags []]
    (if (empty? tokens)
      bags
      (let [color (make-color tokens 1)
            number (Integer/parseInt (nth tokens 0))]
        (recur (drop 4 tokens)
               (conj bags (bag color number [])))))))

(defn line-to-bag
  [line]
  (let [tokens (tokenize line)
        color (make-color tokens 0)
        s (nth tokens 4)
        cnt (if (= "no" s) 0 (Integer/parseInt s))
        contained (if (zero? cnt) [] (make-contained-bags (drop 4 tokens)))]
      (bag color cnt contained)))

(defn find-bag-in-bags-map
  [bags-map bag]
  ((keyword (:color bag)) bags-map))

(defn contains-shiny-gold?
  [bags-map bag]
  (if (some #{"shiny gold"} (map :color (:contained bag)))
    bag
    (let [contained-bags (map (partial find-bag-in-bags-map bags-map) (:contained bag))]
      (first (filter (partial contains-shiny-gold? bags-map) contained-bags)))))
  


(let [bags (mapv line-to-bag (lines "input-day7.txt"))
      bags-map (into {} (mapv 
                          (fn [bag] [(keyword (:color bag)) bag])
                          bags))]
  (println (count (filter (partial contains-shiny-gold? bags-map) bags))))
